tools/f2fsslower.py

```diff
@@ -0,0 +1,338 @@
+    char file[DNAME_INLINE_LEN];
+};
+BPF_HASH(entryinfo, u64, struct val_t);
+PERF_TABLE
+//
+// Store timestamp and size on entry
+//
+// The current f2fs (Linux 4.5) uses generic_file_read_iter(), instead of it's
+// own function, for reads. So we need to trace that and then filter on f2fs,
+// which I do by checking file->f_op.
+// The new Linux version (since form 4.10) uses f2fs_file_read_iter(), And if
+// the 'CONFIG_FS_DAX' is not set, then f2fs_file_read_iter() will call
+// generic_file_read_iter(), else it will call f2fs_dax_read_iter(), and trace
+// generic_file_read_iter() will fail.
+int trace_read_entry(struct pt_regs *ctx, struct kiocb *iocb)
+{
+    u64 id =  bpf_get_current_pid_tgid();
+    u32 pid = id >> 32; // PID is higher part
+    if (FILTER_PID)
+        return 0;
+    // f2fs filter on file->f_op == f2fs_file_operations
+    struct file *fp = iocb->ki_filp;
+    if ((u64)fp->f_op != F2FS_FILE_OPERATIONS)
+        return 0;
+    // store filep and timestamp by id
+    struct val_t val = {};
+    val.ts = bpf_ktime_get_ns();
+    val.fp = fp;
+    val.offset = iocb->ki_pos;
+    if (val.fp)
+        entryinfo.update(&id, &val);
+    return 0;
+}
+// f2fs_file_write_iter():
+int trace_write_entry(struct pt_regs *ctx, struct kiocb *iocb)
+{
+    u64 id = bpf_get_current_pid_tgid();
+    u32 pid = id >> 32; // PID is higher part
+    if (FILTER_PID)
+        return 0;
+    // store filep and timestamp by id
+    struct val_t val = {};
+    val.ts = bpf_ktime_get_ns();
+    val.fp = iocb->ki_filp;
+    val.offset = iocb->ki_pos;
+    if (val.fp)
+        entryinfo.update(&id, &val);
+    return 0;
+}
+// f2fs_file_open():
+int trace_open_entry(struct pt_regs *ctx, struct inode *inode,
+    struct file *file)
+{
+    u64 id = bpf_get_current_pid_tgid();
+    u32 pid = id >> 32; // PID is higher part
+    if (FILTER_PID)
+        return 0;
+    // store filep and timestamp by id
+    struct val_t val = {};
+    val.ts = bpf_ktime_get_ns();
+    val.fp = file;
+    val.offset = 0;
+    if (val.fp)
+        entryinfo.update(&id, &val);
+    return 0;
+}
+// f2fs_sync_file():
+int trace_fsync_entry(struct pt_regs *ctx, struct file *file)
+{
+    u64 id = bpf_get_current_pid_tgid();
+    u32 pid = id >> 32; // PID is higher part
+    if (FILTER_PID)
+        return 0;
+    // store filep and timestamp by id
+    struct val_t val = {};
+    val.ts = bpf_ktime_get_ns();
+    val.fp = file;
+    val.offset = 0;
+    if (val.fp)
+        entryinfo.update(&id, &val);
+    return 0;
+}
+//
+// Output
+//
+static int trace_return(struct pt_regs *ctx, int type)
+{
+    struct val_t *valp;
+    u64 id = bpf_get_current_pid_tgid();
+    u32 pid = id >> 32; // PID is higher part
+    valp = entryinfo.lookup(&id);
+    if (valp == 0) {
+        // missed tracing issue or filtered
+        return 0;
+    }
+    // calculate delta
+    u64 ts = bpf_ktime_get_ns();
+    u64 delta_us = (ts - valp->ts) / 1000;
+    entryinfo.delete(&id);
```

Comment:
Do NOT delete the element prematurely. This should be done after perf submit calls.